# CLas_Phylogenomics
Collection of scripts for a range of genomic data processing and analysis.
Not everything is documented yet, but most scripts have some helpful information if you type `python script.py -h`.

## Contents

* [Download Genbank assemblies](#Download-Genbank-assemblies)
* [Listing SNPs impacted by recombination](#Listing-SNPs-impacted-by-recombination)
* [GC skew](#GC-skew)

## Download Genbank assemblies
The script `GenBankScan.py` automatically downloads [Genbank](https://www.ncbi.nlm.nih.gov/genbank/) assemblies for a given search term (e.g. a species name).

#### Example command
`python GenBankScan.py --term 'Candidatus Liberibacter asiaticus' --format 'GenBank' --email 'other@example.com'`

`python GenBankScan.py -h` Will print a full list of command arguments.

#### Command arguments
| Name | Description |
| :--: | ----------- | 
| `term` | Search term, usually organism name (required) |
| `format`  | Accession format (RefSeq or GenBank) (required) |
| `email`  | Provide your own email address here (required) |

#### Notes
To run it, ensure that you are using Python v.3.7, and have installed the following dependencies: [os](https://docs.python.org/3/library/os.html), [Bio](https://biopython.org/), [urllib](https://docs.python.org/3/library/urllib.html), [argparse](https://docs.python.org/3/library/argparse.html), and [sys](https://docs.python.org/3/library/sys.html).

## Listing SNPs impacted by recombination
The script `ExclRecPos.py` creates a file listing each SNP impacted by recombination. It uses the `importation_status.txt` files generated by [ClonalFrameML](https://github.com/xavierdidelot/ClonalFrameML), which gives the genomic regions impacted by recombination.

#### Example command
`python ExclRecPos.py --inputfile '/path/to/file.txt' --chrom "chrom1"`

`python ExclRecPos.py -h` Will print a full list of command arguments.

#### Command arguments
| Name | Description |
| :--: | ----------- | 
| `inputfile` | Path to input file (required) |
| `chrom`  | Chromosome name (required) |

#### Notes
To run it, ensure that you are using Python v.3.7, and have installed the following dependencies: [os](https://docs.python.org/3/library/os.html), [panda](https://pandas.pydata.org/),[argparse](https://docs.python.org/3/library/argparse.html), [re](https://docs.python.org/3/library/re.html), and [sys](https://docs.python.org/3/library/sys.html).

## GC skew
The script `GCskew.py` calculates the incremental and the cumulative GC skew for a given FASTA sequence according to this formula: Skew = (G - C) / (G + C).

#### Example command
`python GCskew.py --sequence '/path/to/sequence.fasta' --window 100`

`python GCskew.py -h` Will print a full list of command arguments.

#### Command arguments
| Name | Description |
| :--: | ----------- | 
| `sequence` | Path to the FASTA sequence (required) |
| `window`  | Window size (required) |

#### Notes
To run it, ensure that you are using Python v.3.7, and have installed the following dependencies: [os](https://docs.python.org/3/library/os.html), [panda](https://pandas.pydata.org/), [Bio](https://biopython.org/),[argparse](https://docs.python.org/3/library/argparse.html), and [sys](https://docs.python.org/3/library/sys.html).

___
## Citation

Please cite this when using these scripts:
Pruvost O., Boyer K., Labbé F., Weishaar M., Vynisale A., Melot C., Hoareau C., Cellier G., and Ravigné V., 2024. Genetic signatures of contrasted outbreak histories of ‘*Candidatus* Liberibacter asiaticus’, the bacterium that causes citrus huanglongbing in three outermost regions of the European Union. (*in prep*)
